oci/oci.go

```diff
@@ -604,6 +604,53 @@ func waitContainerStop(ctx context.Context, c *Container, timeout time.Duration)
 	return nil
 }
 
+func (r *Runtime) waitContainerStateStopped(ctx context.Context, c *Container, timeout time.Duration) (err error) {
+	done := make(chan error)
+	// we could potentially re-use "done" channel to exit the loop on timeout
+	// but we use another channel "chControl" so that we won't never incur in the
+	// case the "done" channel is closed in the "default" select case and we also
+	// reach the timeout in the select below. If that happens we could raise
+	// a panic closing a closed channel so better be safe and use another new
+	// channel just to control the loop.
+	chControl := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-chControl:
+				return
+			default:
+				// Check if the container is stopped
+				if err := r.UnlockedUpdateStatus(c); err != nil {
```

Comment:
Actually thinking about this more, we are now switching to calling runtime state for each stop operation in a loop which might be an impact to performance. Can we instead get a notification from the watcher?

Reply:
The corner case I can see with adding a channel would be (but correct me if I am wrong):
- The container process returns while CRI-O is called to stop the container
- The watcher catches the notification and call into `UpdateStatus()`, while at the same time, the new channel is registered.
- `StopContainer()` waits on this channel. But nothing will never happen since the event happened already.

@mrunalp Do you think this is a valid case ?

Reply:
@sboeuf yes, that is a valid race.

Reply:
@sboeuf One more option could be letting watcher do what it does today (update status) and you just poll the in-memory state here.

Reply:
> `StopContainer()` waits on this channel. But nothing will never happen since the event happened already.

You can guard against this by waiting until you're attached to the channel, and then (once) checking the current state.  No need to poll after that.

Reply:
This would not work because of the way the watcher is designed. It will see one event and if we miss this event, from the main thread, we'll have no other chance to get the status updated.
I have submitted a similar PR to address @mrunalp concerns here: https://github.com/kubernetes-incubator/cri-o/pull/1334